### PR TITLE
VMware: Add support for more than 15 disks

### LINKED ISSUE #22808: vmware_guest - Won't allow more than 15 disks
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

vmware_guest

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.6.6 (r266:84292, Aug  9 2016, 06:11:56) [GCC 4.4.7 20120313 (Red Hat 4.4.7-17)]
```

##### OS / ENVIRONMENT

RedHat 6.3/7.2
##### SUMMARY

Cannot create VM with more than 15 disks using vmware_guest.  This seems to have to do with the fact that you cannot only have 15 disks on each scsi bus.  This fails with the error: "A specified parameter was not correct: unitNumber"

In looking through the vmware_guest.py script, it appears there's no mechanism to increment the busNumber.

    @staticmethod
    def create_scsi_controller(scsi_type):
        scsi_ctl = vim.vm.device.VirtualDeviceSpec()
        scsi_ctl.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
        if scsi_type == 'lsilogic':
            scsi_ctl.device = vim.vm.device.VirtualLsiLogicController()
        elif scsi_type == 'paravirtual':
            scsi_ctl.device = vim.vm.device.ParaVirtualSCSIController()
        elif scsi_type == 'buslogic':
            scsi_ctl.device = vim.vm.device.VirtualBusLogicController()
        elif scsi_type == 'lsilogicsas':
            scsi_ctl.device = vim.vm.device.VirtualLsiLogicSASController()

        scsi_ctl.device.deviceInfo = vim.Description()
        scsi_ctl.device.slotInfo = vim.vm.device.VirtualDevice.PciBusSlotInfo()
        scsi_ctl.device.slotInfo.pciSlotNumber = 16
        scsi_ctl.device.controllerKey = 100
        scsi_ctl.device.unitNumber = 3
        scsi_ctl.device.busNumber = 0
        scsi_ctl.device.hotAddRemove = True
        scsi_ctl.device.sharedBus = 'noSharing'
        scsi_ctl.device.scsiCtlrUnitNumber = 7

        return scsi_ctl

##### STEPS TO REPRODUCE

```yaml
---
- name: Create VM
  vmware_guest:
    hostname: "{{vsphere_hostname}}"
    username: "{{vsphere_username}}"
    password: "{{vsphere_password}}"
    validate_certs: no
    datacenter: "{{datacenter}}"
    cluster: "{{cluster}}"
    folder: "{{folder}}"
    name: "{{ site_code }}{{server_name | default('test2') }}"
    state: "poweredoff"
    guest_id: rhel7_64Guest
    hardware:
      memory_mb: "{{memory}}"
      num_cpus: "{{cpu}}"
      scsi: paravirtual
    networks:
      - name: "ACCESS"
        device_type: "vmxnet3"
    disk:
      - size_gb: "{{disk1_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk2_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk3_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"  
      - size_gb: "{{disk4_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk5_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk6_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk7_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk8_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk9_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk10_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk11_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk12_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk13_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk14_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk15_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk16_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk17_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk18_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk19_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"
      - size_gb: "{{disk20_size}}"
        type: "thin"
        #autoselect_datastore: True
        datastore: "{{datastore}}"                                                                                                                                          
    wait_for_ip_address: no
```

##### EXPECTED RESULTS

VM is built with multiple disks
##### ACTUAL RESULTS

```
No config file found; using defaults
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available

Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.6/site-packages/ansible-2.3.0.0-py2.6.egg/ansible/plugins/callback/__init__.pyc

PLAYBOOK: site.yml *********************************************************************************************************************************************
1 plays in site.yml

PLAY [127.0.0.1] ***********************************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************
Using module file /usr/lib/python2.6/site-packages/ansible-2.3.0.0-py2.6.egg/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242 `" && echo ansible-tmp-1490042590.13-69315754579242="` echo /root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpnj30Th TO /root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242/ /root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242/setup.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1490042590.13-69315754579242/" > /dev/null 2>&1 && sleep 0'
ok: [127.0.0.1]
META: ran handlers

TASK [estem2 : Create VM] **************************************************************************************************************************************
task path: /opt/ansible/ansible-2.3.0.0/plays/roles/estem2/tasks/main.yml:2
Using module file /usr/lib/python2.6/site-packages/ansible-2.3.0.0-py2.6.egg/ansible/modules/cloud/vmware/vmware_guest.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941 `" && echo ansible-tmp-1490042591.23-130289669432941="` echo /root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpnj30Th TO /root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941/vmware_guest.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941/ /root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941/vmware_guest.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941/vmware_guest.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1490042591.23-130289669432941/" > /dev/null 2>&1 && sleep 0'
fatal: [127.0.0.1]: FAILED! => {
    "changed": true, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "annotation": null, 
            "cluster": "shared_storage", 
            "customization": {}, 
            "customvalues": [], 
            "datacenter": "Staging", 
            "disk": [
                {
                    "datastore": "Deployment1", 
                    "size_gb": "135", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "40", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "40", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "29", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "29", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "81", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "81", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "81", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "81", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "58", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "200", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "50", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "50", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "50", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "50", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "30", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "30", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "58", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "58", 
                    "type": "thin"
                }, 
                {
                    "datastore": "Deployment1", 
                    "size_gb": "50", 
                    "type": "thin"
                }
            ], 
            "esxi_hostname": null, 
            "folder": "/vm/test", 
            "force": false, 
            "guest_id": "rhel7_64Guest", 
            "hardware": {
                "memory_mb": "8192", 
                "num_cpus": "2", 
                "scsi": "paravirtual"
            }, 
            "hostname": "192.168.0.100", 
            "is_template": false, 
            "name": "testserver2", 
            "name_match": "first", 
            "networks": [
                {
                    "device_type": "vmxnet3", 
                    "name": "ACCESS"
                }
            ], 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "resource_pool": null, 
            "state": "poweredoff", 
            "template_src": null, 
            "username": "administrator@vsphere.local", 
            "uuid": null, 
            "validate_certs": false, 
            "wait_for_ip_address": false
        }
    }, 
    "msg": "A specified parameter was not correct: unitNumber"
}
        to retry, use: --limit @/opt/ansible/ansible-2.3.0.0/plays/site.retry

PLAY RECAP *****************************************************************************************************************************************************
127.0.0.1                  : ok=1    changed=0    unreachable=0    failed=1 
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
This fix adds support for adding more than 15 disk to a given
virtual machine. Also, adds integration tests for this change.

Fixes: #22808

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
test/integration/targets/vmware_guest/tasks/create_multi_disks_d1_c1_f0.yml
lib/ansible/modules/cloud/vmware/vmware_guest.py

##### ANSIBLE VERSION

```
2.5devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @pdellaert @angystardust @aperigault

**Comment by BDragonReborn** [comment]
This is still an issue with Ansible 2.5

```yaml
tasks:
    - name: Deploy guest from template
      vmware_guest:
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        state: "{{ vm_state }}"
        validate_certs: "{{ validate_certs }}"
        name: "{{ guest_name }}"
        template: "{{ template }}"
        cluster: "{{ cluster }}"
        resource_pool: "{{ resource_pool }}"
        datacenter: "{{ datacenter }}"
        annotation: "{{ vm_description }}"
        disk:
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_one_size }}"
          datastore: "{{ datastore_one }}"
          type: "{{ datastore_type }}" 
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_two_size }}"
          datastore: "{{ datastore_two }}"
          type: "{{ datastore_type }}" 
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_three_size }}"
          datastore: "{{ datastore_three }}"
          type: "{{ datastore_type }}" 
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_four_size }}" 
          datastore: "{{ datastore_four }}"
          type: "{{ datastore_type }}" 
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_five_size }}"
          datastore: "{{ datastore_five }}"
          type: "{{ datastore_type }}" 
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_six_size }}" 
          datastore: "{{ datastore_six }}"
          type: "{{ datastore_type }}" 
        - autoselect_datastore: "False"
          size_gb: "{{ datastore_seven_size }}" 
          datastore: "{{ datastore_seven }}"
          type: "{{ datastore_type }}" 
        hardware:
          memory_mb: "{{ memory_size }}"
          num_cpus: "{{ cpu_numbers }}"
          scsi: paravirtual
        networks:
        - name: "{{ network_name }}"
          ip: "{{ ip_address }}"
          netmask: "{{ netmask }}"
          gateway: "{{ default_gateway }}"
          domain: "{{ domain_name }}"
          dns_servers:
          - "{{ dns_server_one }}"
          - "{{ dns_server_two }}"
        customization:
          autologin: True
          username: "{{ local_username }}"
          password: "{{ local_password }}"
          hostname: "{{ guest_name }}"
          dns_servers:
          - "{{ dns_server_one }}"
          - "{{ dns_server_two }}"
 ```
       
This works with everything but assigning different datastores to the disks

**Comment by Akasurde** [comment]
@BDragonReborn Are you using latest changes using this PR ? I see only 7 disks in configuration. This change need more than 15 disks. Also, multi-datastore PR is not merged, yet.

**Comment by wfharrigan72** [comment]
Have tested this fix on several system builds.  It works with no issues that I could find.

**Comment by thunderpants73** [comment]
Any idea on which release this will be added in?  We've been using this PR in our current Ansible build and it seems to work as expected.

**Comment by Akasurde** [comment]
@thunderpants73 I am waiting for code review for this PR. Once it is done we can merge this.

**Comment by Akasurde** [comment]
cc @pdellaert @tchernomax Could you please review this ?

**Review by pdellaert — APPROVED** [review]
LGTM, have not tested this on my setup in person, but the code itself looks like it should do what is expected :)

**Comment by Akasurde** [comment]
cc @dericcrago

File: test/integration/targets/vmware_guest/tasks/create_multi_disks_d1_c1_f0.yml
Comment (dericcrago) [inline test/integration/targets/vmware_guest/tasks/create_multi_disks_d1_c1_f0.yml:62]: assert that changes were *not* made, right?

File: test/integration/targets/vmware_guest/tasks/create_multi_disks_d1_c1_f0.yml
Comment (dericcrago) [inline test/integration/targets/vmware_guest/tasks/create_multi_disks_d1_c1_f0.yml:62]: same as earlier comment, about changes *not* being made, this looks to be in a couple of places.

**Review by dericcrago — CHANGES_REQUESTED** [review]
Would you be able to add any tests that modify the disk settings and take a VM from <15 disks to >15 disks to add a controller during reconfigure?

**Comment by dericcrago** [comment]
Hi @Akasurde - I'd still like to see a reconfigure test that crosses the threshold to require adding a new controller, like going from <15 to >15, etc.

**Comment by Akasurde** [comment]
@Tomorrow9 Would you be interested in taking over this PR? Let me know, Thanks.

**Comment by Akasurde** [comment]
superseded by https://github.com/ansible/ansible/pull/66057

**Pull request closed without merging**
